python/ray/experimental/state/api.py

```diff
@@ -1,195 +1,239 @@
+    ) -> Union[Dict, List]:
+        """List resources states
+
+        Args:
+            resource_name: Resource names, i.e. 'jobs', 'actors', 'nodes',
+                see `StateResource` for details.
+            options: List options. See `ListApiOptions` for details.
+            _explain: Print the API information such as API
+                latency or failed query information.
+
+        Returns:
+            A list of queried result from `ListApiResponse`,
+
+        Raises:
+            This doesn't catch any exceptions raised when the underlying request
+            call raises exceptions. For example, it could raise `requests.Timeout`
+            when timeout occurs.
+
+        """
+        # Append the ListApiOptions to the url
+        endpoint = f"/api/v0/{resource.value}?{self._get_query_string(options)}"
+
+        # We don't use `asdict` to avoid deepcopy.
+        # https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict
+        params = {
+            "limit": options.limit,
+            "timeout": options.timeout,
+            "filter_keys": [],
+            "filter_values": [],
+        }
+        for filter in options.filters:
+            filter_k, filter_val = filter
+            params["filter_keys"].append(filter_k)
+            params["filter_values"].append(filter_val)
+
+        # NOTE(rickyyx)
+        # This might raise `requests.Timeout` exceptions and other related requests
+        # exceptions. Users of the functions are expected to handle this.
+        response = self._do_request(
+            "GET",
+            endpoint,
+            timeout=options.timeout,
+            params=params,
+        )
+
+        if response.status_code != 200:
+            self._raise_error(response)
+
+        response = response.json()
+        if response["result"] is False:
+            raise RayStateApiException(
+                "API server internal error. See dashboard.log file for more details. "
+                f"Error: {response['msg']}"
+            )
 
-    if _explain:
         # Print warnings if anything was given.
-        warning_msg = response["data"].get("partial_failure_warning", None)
-        if warning_msg:
-            warnings.warn(warning_msg, RuntimeWarning)
+        warning_msgs = response["data"].get("partial_failure_warning", None)
+        if warning_msgs and _explain:
+            warnings.warn(warning_msgs, RuntimeWarning)
+
+        return response["data"]["result"]
+
+
+"""
+Convenient methods for list_<RESOURCE>
 
-    return r.json()["data"]["result"]
+Supported arguments to the below methods, see `ListApiOptions`:
+    address: The address of the Ray state server. If None, it assumes a running Ray
+        deployment exists and will query the GCS for auto-configuration.
+    timeout: Time for the request.
+    limit: Limit of entries in the result
+"""
 
 
 def list_actors(
-    api_server_url: str = None,
+    address: Optional[str] = None,
```

Comment:
To reduce code dup, we could also opt in for a `**kwarg` based interface:

```
def list_actors(
    address: Optional[str] = None,
    _explain: bool = False, 
    **kwargs
):

  options = ListApiOptions(**kwargs) 
```

Reply:
Right - searched online and seems rather manual. So I feel we could keep this more verbose approach for better maintainability and clarity. 

But I wonder if you have considered such API:

```
def list(
  resource: ....
  limit: ...
  filters: ...
```

So instead of calling `state.api.list_actors(...)`, people now do `state.api.list("actors", ...)`, this is probably fine since I don't foresee very different set of options for different listing right?